### PR TITLE
Add support for distributed training

### PR DESCRIPTION
**Pull request by laurenyu**
*Description of changes:*
Add a class for users to use to set up a parameter server like before to use with distributed training.

A few note:
* this doesn't work if you have only one host but still want the parameter server. Likely an issue with a process ending before it should, but I'm not sure yet. will investigate further.
* naming is hard. feel free to suggest alternatives :)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/resources/mnist/mnist_script_mode.py [hunk 1]
```diff
@@ -114,13 +117,15 @@ def save(model_dir, model):
...
-    num_gpus = int(os.environ['SM_NUM_GPUS'])
+    parser.add_argument('--hosts', type=str, default=os.environ['SM_HOSTS'])
 
     args = parser.parse_args()
 
-    model = train(args.batch_size, args.epochs, args.learning_rate, num_gpus, args.train, args.test,
-                  ast.literal_eval(args.hosts), args.current_host)
-    save(args.model_dir, model)
+    num_gpus = int(os.environ['SM_NUM_GPUS'])
+    hosts = ast.literal_eval(args.hosts)
```
Comment (jesterhazy) [inline test/resources/mnist/mnist_script_mode.py]: remove ast.literal_eval. the env var default is supposed to be json so we should only parse it as json (and throw error if it isn't). and user entering command line should be able to enter repeated values not handwritten json.

File: test/unit/test_distributed.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    'DMLC_NUM_SERVER': str(len(MULTIPLE_HOSTS_LIST)),
+    'DMLC_PS_ROOT_URI': IP_ADDRESS,
+    'DMLC_PS_ROOT_PORT': DEFAULT_PORT,
+    'PS_VERBOSE': DEFAULT_VERBOSITY,
+}
+
+MXNET_COMMAND = "python -c 'import mxnet'"
+
+
+def test_init():
```
Comment (jesterhazy) [inline test/unit/test_distributed.py]: i don't see analogous test for multiple hosts case

File: test/resources/mnist/mnist_script_mode.py (hunk 1 above)
Comment (laurenyu) [reply test/resources/mnist/mnist_script_mode.py]: fixed

File: test/unit/test_distributed.py (hunk 2 above)
Comment (laurenyu) [reply test/unit/test_distributed.py]: added

File: src/sagemaker_mxnet_container/distributed.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    ROLES = ['worker', 'scheduler', 'server']
+
+    def __init__(self, hosts, ps_port='8000', ps_verbose='0'):
+        self.scheduler = self.scheduler_host(hosts)
+        self.hosts = hosts
+        self.ps_port = ps_port
+        self.ps_verbose = ps_verbose
+
+    def scheduler_host(self, hosts):
+        return sorted(hosts)[0]
```
Comment (icywang86rui) [inline src/sagemaker_mxnet_container/distributed.py]: is it necessary to sort this? can we just just the first one? Do all the hosts get this list in the same order?

Comment (icywang86rui) [reply src/sagemaker_mxnet_container/distributed.py]: Also this function does not have unit tests. I think we should tests all public functions.

Comment (laurenyu) [reply src/sagemaker_mxnet_container/distributed.py]: it's probably not necessary, but the user could theoretically override the environment variable and give the list in a different order each time?

added a test. also made the function private because I remembered that I wanted it to be "memoized" (it's not proper memoization, I know, I know...) in `self.scheduler` anyway.

Comment (icywang86rui) [reply src/sagemaker_mxnet_container/distributed.py]: i guess nobody is training with millions of hosts. The over head of sorting it is probably ok.
